### DISCUSSION TITLE
Constructing ad-hoc FieldLayout from config vs referencing saved FieldLayout records

### DISCUSSION THREAD

**Discussion by michaelrog** [craftcms/cms#11899] [Q&A] [answered]
Instantiating ad-hoc field layouts via [`FieldLayout::createFromConfig()`](https://docs.craftcms.com/api/v4/craft-models-fieldlayout.html#method-createfromconfig) is super easy.

And, conveniently the field layout editor submits its data already formatted as a 'config' blob.

With that in mind — if I'm developing a custom component that provides a field layout — **Is it a _bad_ practice to just store a config in the SavableComponent `settings` and construct the `FieldLayout` on-demand from that saved config?**

As far as I know, all the native components take care to save the FieldLayout record(s) separately in the db, reference them by ID, and handle updating them through Project Config. _BUT..._ It seems like I can get away with _not_ going through that more involved process, and everything still works fine.

The possible downsides I can imagine, using this ad-hoc approach, include:
- I don't get help from the db (fks) to automatically remove a field from extant layouts upon deletion.
- The layout config looks a lot messier stored as a giant associative array in Project Config, as opposed to the nicely-formatted structure of other natively-saved layouts.

However,  all these downsides seem more aesthetic than functional, at least on first pass. For example, the FieldLayout will calmly skip over any field that doesn't actually exist (e.g. on that has been deleted since the config was saved). And, other savable components include big ugly associative arrays in their Project Config sometimes, so is it really that bad?

- **Does this ad-hoc approach actually work as trivially as I think it does?**
- **Is this an intended/advisable option for implementation?**
- **What am I missing? (especially gotchas I'm likely to encounter further down the road)**

**Comment by brandonkelly** [comment, marked as answer]
Yeah should be totally fine.

> I don't get help from the db (fks) to automatically remove a field from extant layouts upon deletion.

The `fieldlayoutfields` table is hardly even used anymore. Really just there to aid older field layouts that haven’t been saved since before Craft 3.5. So this isn’t a big deal.

**Reply by michaelrog** [reply]
Gotcha — thanks!

I _have_ noticed that [FieldLayout::createFromConfig()](https://docs.craftcms.com/api/v4/craft-models-fieldlayout.html#method-createfromconfig) ignores the `type` property in the config, so I have to do some weird rigmarole to make sure the correct `type` is attached — Otherwise, my element's native fields won't be included...

For example, this _doesn't_ work:
```
$fieldLayout = FieldLayout::createFromConfig($config);

// $fieldLayout->type is null  :(
```

...but this does:

```
$tabsFromConfig = FieldLayout::createFromConfig($config)->getTabs();
/*
 * `createFromConfig` doesn't set the type on the FieldLayout, so native fields won't be included initially.
 * We need to re-run `setTabs` after setting the type, to include native fields.
 */
$fieldLayout = new FieldLayout([
    'type' => Link::class,
]);
$fieldLayout->setTabs($tabsFromConfig);

// $fieldLayout->type is correct  :/
```

Is that intended behavior?

**Reply by brandonkelly** [reply]
Have you tried this?

```php
$fieldLayout = FieldLayout::createFromConfig($config);
$fieldLayout->type = Link::class;
```

**Reply by michaelrog** [reply]
Yeah — but `EVENT_DEFINE_NATIVE_FIELDS` gets fired from inside of `setTabs()`, so assigning the `type` _after_ `createFromConfig()` doesn't currently work for registering native fields.

**Reply by brandonkelly** [reply]
Ah, ok. This should work then:

```php
$config = ['type' => Link::class] + $config;
$fieldLayout = FieldLayout::createFromConfig($config);
```

**Reply by michaelrog** [reply]
> Ah, ok. This should work then:
> 
> ```
> $config = ['type' => Link::class] + $config;
> $fieldLayout = FieldLayout::createFromConfig($config);
> ```

Great — That's how I'd expect it to work, also. Alas, as currently implemented, `createFromConfig()` instantiates tabs using `$config['tabs']` and _ignores all other properties in the `$config`_.

Created an issue: https://github.com/craftcms/cms/issues/11929

I'll try to pop over a PR to fix that. 👍🏼

**Reply by michaelrog** [reply]
heh you're too fast. 😅  Thank you!!!
